File: adapters/claude/src/agent_wake_claude/reply.py
"""Reply tool for agent-wake Claude adapter.

Spec reference: v1-daemon-spec.md §9.4.

No longer makes HTTPS calls directly.  Sends a reply frame to the
daemon and awaits the matching reply_result via ReplyResultBus.
"""

import uuid

from . import client as daemon_client


def get_tool_definition() -> dict:
    return {
        "name": "agent_wake_reply",
        "description": "Reply to an external event source via agent-wake.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "source": {"type": "string"},
                "content": {"type": "string"},
                "in_reply_to": {"type": "string"},
            },
            "required": ["source", "content"],
        },
    }


def handle_reply_tool_call(arguments: dict) -> dict:
    source = arguments.get("source")
    content = arguments.get("content")
    in_reply_to = arguments.get("in_reply_to")

    if not source or not content:
        return {
            "content": [
                {"type": "text", "text": "missing required fields: source, content"}
            ],
        }

    reply_id = str(uuid.uuid4())
    frame = {
        "type": "reply",
        "reply_id": reply_id,
        "source": source,
        "in_reply_to": in_reply_to or "",
        "content": content,
    }

    try:
        daemon_client.send_reply_frame(frame)
    except Exception as e:
        return {
            "content": [
                {"type": "text", "text": f"reply delivery failed: {e}"}
            ],
            "isError": True,
        }

    result = daemon_client.ReplyResultBus.wait_result(reply_id, timeout=35.0)
    if result is None:
        return {
            "content": [
                {"type": "text", "text": "reply delivery failed: timed out"}
            ],
            "isError": True,
        }

    status = result.get("status")
    if status == "delivered":
        return {"content": [{"type": "text", "text": "sent"}]}
    if status == "no_callback":
        return {
            "content": [
                {"type": "text", "text": "sent (no callback_url configured)"}
            ],
        }
    error = result.get("error", "unknown error")
    return {
        "content": [{"type": "text", "text": f"reply delivery failed: {error}"}],
        "isError": True,
    }

File: adapters/opencode/src/reply.ts
/**
 * Reply tool: agent calls this to send a message back to the originating
 * event source. The tool sends a `reply` frame to the daemon and awaits
 * the matching `reply_result`. The daemon owns the outbound HTTPS POST.
 */
import { randomUUID } from "node:crypto";
import { ReplyResultBus, sendReplyFrame, type ReplyResult } from "./client";

export interface ReplyArgs {
  source: string;
  content: string;
  in_reply_to?: string;
}

const MAX_REPLY_SIZE = 65536; // 64 KiB

export async function executeReply(args: ReplyArgs): Promise<string> {
  if (!args.source || !args.content) {
    return "missing required fields: source, content";
  }

  let content = args.content;
  if (new TextEncoder().encode(content).length > MAX_REPLY_SIZE) {
    const bytes = new TextEncoder().encode(content);
    content = new TextDecoder("utf-8", { fatal: false }).decode(bytes.slice(0, MAX_REPLY_SIZE));
  }

  const replyId = randomUUID();
  const pending = ReplyResultBus.register(replyId, 35000);

  try {
    sendReplyFrame({
      type: "reply",
      reply_id: replyId,
      source: args.source,
      in_reply_to: args.in_reply_to ?? "",
      content: content,
    });
  } catch (e: any) {
    ReplyResultBus.deliver({
      reply_id: replyId,
      status: "failed",
      error: String(e?.message ?? e),
    });
    return `reply delivery failed: ${e?.message ?? e}`;
  }

  let result: ReplyResult;
  try {
    result = await pending;
  } catch (e: any) {
    return `reply delivery failed: ${e?.message ?? e}`;
  }

  if (result.status === "delivered") return "sent";
  if (result.status === "no_callback") return "sent (no callback_url configured)";
  return `reply delivery failed: ${result.error ?? "unknown error"}`;
}
